### ISSUE TITLE
Assertion failure when pushing a block while someone was flooding the network

### ISSUE THREAD

**Issue by abitmore** [cryptonomex/graphene#292]
See https://bitsharestalk.org/index.php/topic,17962.msg233598.html#msg233598

```
2015-09-03T00:30:20 th_a:invoke handle_block         handle_block ] Got block #34498 from network           application.cpp:356
...
...
2015-09-03T00:30:30 th_a:invoke handle_transaction   handle_transaction ] Got transaction from network          application.cpp:391
2015-09-03T00:30:30 th_a:invoke handle_transaction   handle_transaction ] Got transaction from network          application.cpp:391
2015-09-03T00:30:30 th_a:invoke handle_block         handle_block ] Got block #34500 from network           application.cpp:356
2015-09-03T00:30:30 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.         fork_database.cpp:57
2015-09-03T00:30:35 th_a:invoke handle_transaction   handle_transaction ] Got transaction from network          application.cpp:391
2015-09-03T00:30:35 th_a:invoke handle_block         handle_block ] Got block #34501 from network           application.cpp:356
2015-09-03T00:30:35 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.         fork_database.cpp:57
2015-09-03T00:30:37 th_a:invoke handle_block         handle_block ] Got block #34499 from network           application.cpp:3
...
...
2015-09-03T00:30:37 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
(skip & skip_transaction_dupe_check) || trx_idx.indices().get<by_trx_id>().find(trx_id) == trx_idx.indices().get<by_trx_id>().end(): 
    {}
    th_a  db_block.cpp:487 _apply_transaction

    {"trx":{"ref_block_num":34497,"ref_block_prefix":2564640669,"expiration":"2015-09-03T00:31:49","operations":[[0,{"fee":{"amount":2000000,"asset_id":"1.3.0"},"from":"1.2.63354","to":"1.2.63355","amount":{"amount":100000,"asset_id":"1.3.0"},"extensions":[]}]],"extensions":[],"signatures":["1f5f41592441c4f40af47b118e80a0c4786a1d768cf6a8d3204d7c30d79d33e971463c4dc78f1c5fc8f2ae7637ec21e8b523b089baef754422fe99108576b343c8"]}}
    th_a  db_block.cpp:543 _apply_transaction

    {"next_block.block_num()":34499}
    th_a  db_block.cpp:448 _apply_block
...
...
2015-09-03T00:30:37 th_a:invoke handle_transaction   handle_transaction ] Got transaction from network          application.cpp:391
2015-09-03T00:30:37 th_a:invoke handle_transaction   on_objects_changed ] ids: ["2.5.40","2.5.18","2.6.63354","2.7.4698"]           api.cpp:867
2015-09-03T00:30:37 th_a:invoke handle_transaction   on_objects_changed ] ids: ["2.5.40","2.5.18","2.6.63354","2.7.4698"]           api.cpp:867
2015-09-03T00:30:40 th_a:invoke handle_block         handle_block ] Got block #34502 from network           application.cpp:356
2015-09-03T00:30:40 th_a:invoke handle_block           push_block ] Pushing block to fork database that failed to link.         fork_database.cpp:57
```

Full default.log: https://drive.google.com/open?id=0B3xrm70jSHn4U3h3V3liOUlOakk
Full p2p.log: https://drive.google.com/open?id=0B3xrm70jSHn4SURQZjBZaE1rVGc

//Edit: link of p2p.log updated, the original is broken

**Comment by abitmore** [comment]
Further more, unable to sync after 'clean' restarting.
See https://bitsharestalk.org/index.php/topic,17962.msg233613.html#msg233613
- The head_block of the chain was block 34498, id = 000086c202f51da427ebf0507404bc9b6d192f96 
- Witness failed to push block 34499 last time, id = 000086c3bf20d2e5eeb1039f38c221b46880f2f0
- After restarted, the witness ignores block 34499 (id 000086c3bf20d2e5eeb1039f38c221b46880f2f0) because it's considered "seen"
- Following blocks are unable to push to chain because they're not linked (because 34499 is not on chain)

```
2015-09-03T09:46:30 p2p:message read_loop fetch_next_batch_of_ ] sync: sending a request for the next items after 000086c202f51da427ebf0507404bc9b6d192f96 to peer 176.9.234.167:40936, (full request is ["000006c280cdf25ca122561981716c74045f16fe","000046c2482fc56dda591b0bea7a47ccbd64d6b2","000066c27fc3abd468f7dd5ad1ce8939614d9d64","000076c2a8ad357d929c0967d98ba955521c11b4","00007ec23da7bcf6188fa3e6e57f0216e78cba1a","000082c22a40c3cbc1152bbd18239e9ffca051dc","000084c22add8beb2d06e3adb13c3573a56ae0b4","000085c29de6e5d439166100ab9d809c730e006c","00008642bd45a1a5711a74bacf4e73cfd5ae69cd","00008682ea2e75d79000e7a4c3411368d7a0a74a","000086a273f0cd7e751e91f6b0795486303009a2","000086b28863612b25b49a1c0a95a3525e42529c","000086ba6b31127d85ba6e1a8b5ea1145a0766e0","000086be74b88f7adb24796112e15c72b04a4e59","000086c0a1c19fb2c9b2d8d92f1caf812fda4bde","000086c19d4fdd989e8e7a62830c6f58e73f33be","000086c202f51da427ebf0507404bc9b6d192f96"])          node.cpp:2294
...
...
2015-09-03T09:46:30 p2p:message read_loop on_blockchain_item_i ] sync: received a list of 669 available items from 192.241.198.6:39809          node.cpp:2310
2015-09-03T09:46:30 p2p:message read_loop on_blockchain_item_i ] is_first_item_for_other_peer: false.  item_hashes_received.size() = 669            node.cpp:2356
2015-09-03T09:46:30 p2p:message read_loop on_blockchain_item_i ] popping item because delegate has already seen it.  peer's last block the delegate has seen is now 000086c202f51da427ebf0507404bc9b6d192f96 (1)            node.cpp:2368
2015-09-03T09:46:30 p2p:message read_loop on_blockchain_item_i ] popping item because delegate has already seen it.  peer's last block the delegate has seen is now 000086c3bf20d2e5eeb1039f38c221b46880f2f0 (2)            node.cpp:2368
2015-09-03T09:46:30 p2p:message read_loop on_blockchain_item_i ] after removing all items we have already seen, item_hashes_received.size() = 667           node.cpp:2371
2015-09-03T09:46:30 p2p:message read_loop trigger_fetch_sync_i ] Triggering fetch sync items loop now           node.cpp:1084
2015-09-03T09:46:30 p2p:fetch_sync_items_loop fetch_sync_items_loo ] beginning another iteration of the sync items loop         node.cpp:1023
2015-09-03T09:46:30 p2p:fetch_sync_items_loop request_sync_items_f ] requesting 100 item(s) ["000086c4e3da8babd770344cb8db18ee89a9d09a","000086c50ebdc7f4b5ff303fee02b53f3aec5773",...,"00008727dc9a24b61d0d51454278f2282e89672d"] from peer 192.241.198.6:39809            node.cpp:1007
...
...
2015-09-03T09:48:39 p2p:send_sync_block_to_node_delegate send_sync_block_to_n ] Successfully pushed sync block 34500 (id:000086c4e3da8babd770344cb8db18ee89a9d09a)          node.cpp:2793
2015-09-03T09:48:39 p2p:send_sync_block_to_node_delegate send_sync_block_to_n ] sync: client accpted the block, we now have only 5937 items left to fetch before we're in sync          node.cpp:2831
```

Full default.log: https://drive.google.com/open?id=0B3xrm70jSHn4UWJ2N2QzdkVOX2M
Full p2p.log: https://drive.google.com/open?id=0B3xrm70jSHn4QVNoanZfMS13UXM

**Comment by abitmore** [comment]
Happened again earlier today.

```
2015-09-05T06:55:06 th_a:invoke handle_block         handle_block ] Got block #67789 from network                       application.cpp:356
2015-09-05T06:55:06  th_a:invoke get_item             get_item ] Request for item {"item_type":1001,"item_hash":"000108cd1efd3b51d003635c35f113a91553a2ab"}                     application.cpp:451
2015-09-05T06:55:06  th_a:invoke get_item             get_item ] Serving up block #67789                        application.cpp:459
2015-09-05T06:55:07 th_a:Witness Block Production block_production_loo ] Not producing block because slot has not yet arrived
                witness.cpp:223
2015-09-05T06:55:07 th_a:Witness Block Production schedule_production_ ] now.time_since_epoch().count(): 1441436106947624 next_second:
 2015-09-05T06:55:08                      witness.cpp:187
2015-09-05T06:55:07 th_a:invoke handle_transaction   handle_transaction ] Got transaction from network                  application.cpp:391
...
...
2015-09-05T06:55:36 th_a:invoke handle_block         handle_block ] Error when pushing block:
10 assert_exception: Assert Exception
(skip & skip_transaction_dupe_check) || trx_idx.indices().get<by_trx_id>().find(trx_id) == trx_idx.indices().get<by_trx_id>().end():
    {}
    th_a  db_block.cpp:487 _apply_transaction

    {"trx":{"ref_block_num":2252,"ref_block_prefix":1464535407,"expiration":"2015-09-05T06:55:20","operations":[[0,{"fee":{"amount":20
89843,"asset_id":"1.3.0"},"from":"1.2.7015","to":"1.2.67159","amount":{"amount":1,"asset_id":"1.3.0"},"memo":{"from":"GPH5MdK4afYd8gAo
gtwYjuRWAHfSpsCYutfBvZ6y2AeZHLAtcYpY6","to":"GPH6BJYGHftujnbttFFKX6YacnvsMd4sbJrbucg682GiU4vmXHTik","nonce":"14917363401952319281","me
ssage":"3235bff7de4ce33ed8517cc3f18b3918"},"extensions":[]}]],"extensions":[],"signatures":["1f11c6e4ccb763263ca91959282ba26bc44e6a011
8b25cb5b18e51f2a2df8f65f62475c45863bf39b3f2fc61fd487e31efe833fac79b6cccedb6658024195d7d64"]}}
    th_a  db_block.cpp:543 _apply_transaction

    {"next_block.block_num()":67790}
    th_a  db_block.cpp:448 _apply_block

    {"new_block":{"previous":"000108cdc3f0a574bf1ac85724e90889dd2a0c81","timestamp":"2015-09-05T06:55:20","witness":"1.6.1063"......}
```

Will upload full log files later.

**Comment by abitmore** [comment]
Restarted with `--replay-blockchain` and got a crash.

```
   94.4106%   64000 of 67789
   97.3609%   66000 of 67789
witness_node: /app/bts/graphene-test2b/libraries/fc/include/fc/optional.hpp:192: T& fc::optional<T>::operator*() [with T = graphene::chain::signed_block]: Assertion `_valid' failed.

Program received signal SIGABRT, Aborted.
0x00007ffff6c01cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56      ../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff6c01cc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff6c050d8 in __GI_abort () at abort.c:89
#2  0x00007ffff6bfab86 in __assert_fail_base (fmt=0x7ffff6d4b830 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=assertion@entry=0x29e827e "_valid",
    file=file@entry=0x29e8240 "/app/bts/graphene-test2b/libraries/fc/include/fc/optional.hpp", line=line@entry=192,
    function=function@entry=0x29ed400 <fc::optional<graphene::chain::signed_block>::operator*()::__PRETTY_FUNCTION__> "T& fc::optional<T>::operator*() [with T = graphene::chain::signed_block]") at assert.c:92
#3  0x00007ffff6bfac32 in __GI___assert_fail (assertion=0x29e827e "_valid",
    file=0x29e8240 "/app/bts/graphene-test2b/libraries/fc/include/fc/optional.hpp", line=192,
    function=0x29ed400 <fc::optional<graphene::chain::signed_block>::operator*()::__PRETTY_FUNCTION__> "T& fc::optional<T>::operator*() [with T = graphene::chain::signed_block]") at assert.c:101
#4  0x0000000001f3d9e8 in fc::optional<graphene::chain::signed_block>::operator* (this=0x7fffffffcb60)
    at /app/bts/graphene-test2b/libraries/fc/include/fc/optional.hpp:192
#5  0x000000000229f327 in graphene::chain::database::reindex (this=0x323f178, data_dir=..., initial_allocation=...)
    at /app/bts/graphene-test2b/libraries/chain/db_management.cpp:58
#6  0x0000000001f2e07c in graphene::app::detail::application_impl::startup (this=0x323f040)
    at /app/bts/graphene-test2b/libraries/app/application.cpp:269
#7  0x0000000001f22962 in graphene::app::application::startup (this=0x7fffffffdd40)
    at /app/bts/graphene-test2b/libraries/app/application.cpp:662
#8  0x0000000001ef904f in main (argc=21, argv=0x7fffffffe1f8) at /app/bts/graphene-test2b/programs/witness_node/main.cpp:155
```

**Comment by abitmore** [comment]
This happened on both of my 2 nodes today.

**Comment by theoreticalbts** [comment]
The crash at least should have been fixed by 70d3f36fba9ff6ac466f117f7fe241d3584029c3

**Comment by abitmore** [comment]
Closing because it's too old.
